docs/unity/part-2b-c-sharp.md

```diff
@@ -0,0 +1,344 @@
+}
+```
+
+Our final reducer handles player movement. In `UpdatePlayerPosition` we look up the `PlayerComponent` using the user's Identity. If we don't find one, we return an error because the client should not be sending moves without calling `CreatePlayer` first.
+
+Using the `EntityId` in the `PlayerComponent` we retrieved, we can lookup the `EntityComponent` that stores the entity's locations in the world. We update the values passed in from the client and call the auto-generated `Update` function.
+
+**Append to the bottom of lib.cs:**
+
+```csharp
+/// Updates the position of a player. This is also called when the player stops moving.
+[SpacetimeDB.Reducer]
+private static void UpdatePlayerPosition(
+   DbEventArgs dbEvent,
+   StdbVector3 position,
+   float direction,
+   bool moving)
+{
+   // First, look up the player using the sender identity
+   PlayerComponent? player = PlayerComponent.FindByIdentity(dbEvent.Sender);
+   if (player is null)
+   {
+       throw new ArgumentException("Player not found");
+   }
+   // Use the Player's EntityId to retrieve and update the EntityComponent
+   ulong playerEntityId = player.EntityId;
+   EntityComponent? entity = EntityComponent.FindByEntityId(playerEntityId);
+   if (entity is null)
+   {
+       throw new ArgumentException($"Player Entity '{playerEntityId}' not found");
+   }
+   
+   entity.Position = position;
+   entity.Direction = direction;
+   entity.Moving = moving;
+   EntityComponent.UpdateByEntityId(playerEntityId, entity);
+}
+```
+
+---
+
+**Server Validation**
+
+In a fully developed game, the server would typically perform server-side validation on player movements to ensure they comply with game boundaries, rules, and mechanics. This validation, which we omit for simplicity in this tutorial, is essential for maintaining game integrity, preventing cheating, and ensuring a fair gaming experience. Remember to incorporate appropriate server-side validation in your game's development to ensure a secure and fair gameplay environment.
+
+---
+
+### Publishing a Module to SpacetimeDB
+
+Now that we've written the code for our server module and reached a clean checkpoint, we need to publish it to SpacetimeDB. This will create the database and call the init reducer. In your terminal or command window, run the following commands.
+
+```bash
+cd server
+spacetime publish -c unity-tutorial
+```
+
+### Finally, Add Chat Support
+
+The client project has a chat window, but so far, all it's used for is the message of the day. We are going to add the ability for players to send chat messages to each other.
+
+First lets add a new `ChatMessage` table to the SpacetimeDB module. Add the following code to ``lib.cs``.
+
+**Append to the bottom of server/src/lib.cs:**
+
+```csharp
+[SpacetimeDB.Table]
+public partial class ChatMessage
+{
+   // The primary key for this table will be auto-incremented
+   [SpacetimeDB.Column(ColumnAttrs.PrimaryKeyAuto)]
```

Comment:
(not flaming Dylan here) This was clearly never tested because it doesn't work and it's a typo, there is supposed to be a field here `public ulong MessageId`.

Reply:
> (not flaming Dylan here) This was clearly never tested because it doesn't work and it's a typo, there is supposed to be a field here `public ulong MessageId`.

It was very much tested; this is just a typo. Reference is [from the PR prereq](https://github.com/clockworklabs/zeke-demo-project/pull/1/files#diff-1ef0f082e5655de40939d4c0b7b9a117db7ee68d577a9c612ac0fa4f329aacd9R212). Specifically, it should be `ChatEntityId`:

![image](https://github.com/user-attachments/assets/fdd8481a-d603-4008-a142-35afe9a76ab9)

Reply:
> > (not flaming Dylan here) This was clearly never tested because it doesn't work and it's a typo, there is supposed to be a field here `public ulong MessageId`.
> 
> Typo just in the docs, but not the implementation it was based upon where it was very much tested (and the field exists). This PR was respectfully merged by two reviewers ~5mo ago that seemed to stumble upon the same mistake of not noticing -> works fine [in the open PR](https://github.com/clockworklabs/zeke-demo-project/pull/1) for the new demo. Here's the [line ref](https://github.com/clockworklabs/zeke-demo-project/pull/1/files#diff-1ef0f082e5655de40939d4c0b7b9a117db7ee68d577a9c612ac0fa4f329aacd9R212) from the PR. Specifically, it should be `ChatEntityId` if aiming to use the new tutorial:
> 
> ![image](https://private-user-images.githubusercontent.com/8840024/368387042-fdd8481a-d603-4008-a142-35afe9a76ab9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY2MjUwMDMsIm5iZiI6MTcyNjYyNDcwMywicGF0aCI6Ii84ODQwMDI0LzM2ODM4NzA0Mi1mZGQ4NDgxYS1kNjAzLTQwMDgtYTE0Mi0zNWFmZTlhNzZhYjkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkxOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MThUMDE1ODIzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2Q5ZDQ1ZmZkM2Y0YmUyMzY4OWQ4YzMyZTg2ZTcyNWFkZWJlZWI1YzNjNTdkZDY2N2VjZTcxOTdhNWViOGQ2MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.pJJ5imjYwsop7uw44134A1Kzfu0HvRWgoPzF6GeXN4w)

Understood, thanks for the context Dylan 👍